src/ess_rhat.jl

```diff
@@ -233,25 +236,206 @@ The ESS and ``\\widehat{R}`` values can be computed for the following estimators
+function _ess(estimator, samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
+    x = _expectand_proxy(estimator, samples)
     if x === nothing
-        throw(ArgumentError("the estimator $f is not yet supported by `ess_rhat`"))
+        throw(ArgumentError("the estimator $estimator is not yet supported by `ess`"))
+    end
+    return _ess(Val(:basic), x; kwargs...)
+end
+function _ess(::Val{T}, ::AbstractArray{<:Union{Missing,Real},3}; kwargs...) where {T}
+    return throw(ArgumentError("the `type` `$T` is not supported by `ess`"))
+end
+function _ess(type::Val{:basic}, samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
+    return first(_ess_rhat(type, samples; kwargs...))
+end
+function _ess(type::Val{:bulk}, samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
+    return first(_ess_rhat(type, samples; kwargs...))
+end
+function _ess(
+    ::Val{:tail},
+    x::AbstractArray{<:Union{Missing,Real},3};
+    tail_prob::Real=1//10,
+    kwargs...,
+)
+    # workaround for https://github.com/JuliaStats/Statistics.jl/issues/136
+    T = Base.promote_eltype(x, tail_prob)
+    pl = convert(T, tail_prob / 2)
+    pu = convert(T, 1 - tail_prob / 2)
+    S_lower = _ess(Base.Fix2(Statistics.quantile, pl), x; kwargs...)
+    S_upper = _ess(Base.Fix2(Statistics.quantile, pu), x; kwargs...)
+    return map(min, S_lower, S_upper)
+end
+function _ess(::Val{:rank}, samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
+    return _ess(Val(:bulk), samples; kwargs...)
+end
+
+"""
+    rhat(samples::AbstractArray{Union{Real,Missing},3}; type=:rank, split_chains=2)
+
+Compute the ``\\widehat{R}`` diagnostics for each parameter in `samples` of shape
+`(chains, draws, parameters)`. [^VehtariGelman2021]
+
+`type` indicates the type of ``\\widehat{R}`` to compute (see below).
+
+$_DOC_SPLIT_CHAINS
+
+See also [`ess`](@ref), [`ess_rhat`](@ref), [`rstar`](@ref)
+
+## Types
+
+The following `type`s are supported:
+- `:rank`: maximum of ``\\widehat{R}`` with `type=:bulk` and `type=:tail`.
+- `:bulk`: basic ``\\widehat{R}``` computed on rank-normalized draws. This type diagnoses
+    poor convergence in the bulk of the distribution due to trends or different locations of
+    the chains.
+- `:tail`: ``\\widehat{R}`` computed on draws folded around the median and then
+    rank-normalized. This type diagnoses poor convergence in the tails of the distribution
+    due to different scales of the chains.
+- `:basic`: Classic ``\\widehat{R}``.
+
+[^VehtariGelman2021]: Vehtari, A., Gelman, A., Simpson, D., Carpenter, B., & Bürkner, P. C. (2021).
+    Rank-normalization, folding, and localization: An improved ``\\widehat {R}`` for
+    assessing convergence of MCMC. Bayesian Analysis.
+    doi: [10.1214/20-BA1221](https://doi.org/10.1214/20-BA1221)
+    arXiv: [1903.08008](https://arxiv.org/abs/1903.08008)
+"""
+@constprop :aggressive function rhat(
+    samples::AbstractArray{<:Union{Missing,Real},3}; type=Val(:rank), kwargs...
```

Comment:
Again could we just support `Symbol`? Or only `Val`? In any case, I think we could add a type restriction to `type`.